openedx/core/djangoapps/catalog/docs/decisions/002-course-models.rst

```diff
@@ -0,0 +1,76 @@
+Course Models
+#############
+
+Status
+******
+
+Accepted
+
+Decision
+********
+
+edx-platform will have a first-class model to represent the idea of a catalog
+course. If we think of multiple runs of the same course to all derive from one
+conceptiual course.  The catalog course would be that conceptual course.
+
+Context
+*******
+
+Historically this concept has been inferred from the names of courses with the
+runs removed (eg. "AximX/Eng101" could be a catalog course with multiple runs
+(2024H1, 2024H2, etc)). Having this explicitly defined as a new model has
+multiple benefits.
+
+* Relationships between course runs can be defined explicitly instead of implicitly.
+
+  * Currently this information is either implied or we require the entire
+    course-discovery service in order to retrieve this information.
+
+* A list of all courses can be easily queried both internally and via the rest API.
+
+* Courses can be explicitly mapped to organizations rather than implicitly.
+
+* This change opens up the door for program or course bundle data to be persisted
+so that a reset of mem cache doesn't break programs related features.
+
+* Org → Course → Course Runs mappings don't have to follow implicit naming
+  Conventions and associations can be changed more easily.
+
+Rejected Alternatives
+*********************
+
+Rely on course-discovery for CatalogCourse
+==========================================
+
+The course-discovery service already has some models to represent this and we
+could rely on it directly but it also contains a lot of representations that
+are specific to the edx.org process and it's not valuable to spin up the
+course-discovery service just to get this information for most operators.
```

Comment:
1. Is there is an existing toggle, or a new toggle, which will be used to choose between different behaviors?
2. Will the implementation be fully backward-compatible? Will the implementation fully support both using and not using course-discovery? For example, if discovery code is needed to call the new APIs, is that intended to be part of this implementation?
3. Is there any longer-term deprecation of discovery in mind that this is a part of? Or possibly just making the external catalog more pluggable?
4. Separate, but related, there are a number of places where the LMS makes synchronous calls to Discovery, which causes reliability issues for the LMS when Discovery is having issues. Do you know if a benefit of this design would be to replace any of these calls, even in cases where Discovery is being used as a separate service?

Reply:
@robrap Can you elaborate on 4?

I know of the the [programs cache](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/catalog/docs/decisions/001-programs-cache.rst), which this will definitely help.

My search just now also introduced me to the [learner_skills_api](https://github.com/openedx/edx-platform/pull/31417), which this will not directly help since it relies on course-discovery's taxonomy API. In the long-term, though, taking course-discovery-querying utils out of edx-platform will have the benefit of discouraging new instances of the edx-platform-queries-course-discovery anti-pattern.

Do you know of other instances?

Reply:
Regarding #4, I haven't actually looked into the problem specifically. I've just seen its affect when we've had temporary issues on discovery that back up into the LMS. I probably need to look more closely when we have an actual incident. Thanks.